registry/machine.go

```diff
@@ -93,9 +106,78 @@ func (r *EtcdRegistry) SetMachineState(ms machine.MachineState, ttl time.Duratio
 		return uint64(0), err
 	}
 
+	// Set the initial metadata on creation
+	if err = r.setMachineMetadata(ms); err != nil {
+		return uint64(0), err
+	}
+
 	return resp.Node.ModifiedIndex, nil
 }
 
+func (r *EtcdRegistry) UpdateMachineMetadata(machID string, key string, value string) error {
+	//Attempt to update key
+	update := etcd.Update{
+		Key:   path.Join(r.keyPrefix, machinePrefix, machID, "metadata", key),
+		Value: value,
+	}
+
+	_, err := r.etcd.Do(&update)
+	if err == nil {
+		return nil
+	}
+
+	// Create
+	create := etcd.Create{
+		Key:   path.Join(r.keyPrefix, machinePrefix, machID, "metadata", key),
+		Value: value,
+	}
+
+	_, err = r.etcd.Do(&create)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *EtcdRegistry) DeleteMachineMetadata(machID string, key string) error {
+	del := etcd.Delete{
+		Key: path.Join(r.keyPrefix, machinePrefix, machID, "metadata", key),
+	}
+
+	_, err := r.etcd.Do(&del)
+	if isKeyNotFound(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Remove and reset all metadata
+func (r *EtcdRegistry) setMachineMetadata(ms machine.MachineState) error {
```

Comment:
Just pass in the fields of MachineState that are relevant to this method: ID and Metadata. It simplifies the testing matrix.